### PR TITLE
Edit boid code to enable easy implementation of new boid types

### PR DESCRIPTION
**Pull request by ahanton**
The boid class is now an abstract parent of the boid type scripts. Each boid type script only implements what is specific to that type.

### CODE REVIEW

File: Assets/Prefabs/Units/Melee.prefab [hunk 1]
```diff
@@ -73,24 +73,24 @@ MonoBehaviour:
...
+  cost: 0
+  health: 0
+  damage: 0
+  maxSpeed: 0
+  targetHeight: 0
+  collisionAvoidanceDistance: 0
+  avoidCollisionWeight: 0
+  hover_Ki: 0
+  hover_Kp: 0
+  timeBetweenAttacks: 0
```
Comment (frans-bergman) [inline Assets/Prefabs/Units/Melee.prefab:88]: Why not set the properties in the prefab?

**Review by frans-bergman — COMMENTED** [review]
Why do you set the properties in `Start` instead of in the prefab?

File: Assets/Scripts/Boid/Melee.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        maxSpeed = 2f;
+        targetHeight = 2f;
+        collisionAvoidanceDistance = 3f;
+        avoidCollisionWeight = 5f;
+        hover_Ki = 2f;
+        hover_Kp = 10f;
+        timeBetweenAttacks = 0.1f;
+        dead = false;
+        //mesh = ;
+        collisionMask = LayerMask.GetMask("Wall", "Obstacle");
```
Comment (frans-bergman) [inline Assets/Scripts/Boid/Melee.cs:26]: I think the layer is called "Walls".

File: Assets/Prefabs/Units/Melee.prefab (hunk 1 above)
Comment (palmdrop) [reply Assets/Prefabs/Units/Melee.prefab:88]: Agree, if it's possible to set in the prefab we should do it. Will make it easier to balance the weights of different units.

**Review by palmdrop — COMMENTED** [review]
Looks good, but would be nice if it was possible to edit the weights using the editor?

**Comment by ahanton** [comment]
I think it is better to have everything in the same place (classInfo can't be set in the prefab) so I removed [SerializeField] for all fields and from now on all values will be set in the script.

**Review by frans-bergman — APPROVED** [review]
Veeery good. :1st_place_medal:

**Review by palmdrop — APPROVED** [review]
Looks good. Doesn't seem to be a way to expose struct fields in the editor, so having all the fields set in the constructor is probably the second-best solution. Approved.

**Pull request merged**